crates/schema-inference/src/inference.rs

```diff
@@ -0,0 +1,178 @@
+use doc::inference::{ArrayShape, ObjProperty, ObjShape, Shape};
+use json::schema::types;
+use serde_json::Value as JsonValue;
+
+use crate::shape;
+
+pub fn infer_shape(value: &JsonValue) -> Shape {
+    match value {
+        JsonValue::Bool(value) => infer_bool(value),
+        JsonValue::Number(value) => infer_number(value),
+        JsonValue::String(value) => infer_string(value.as_ref()),
+        JsonValue::Null => infer_null(),
+        JsonValue::Array(inner) => infer_array(inner.as_slice()),
+        JsonValue::Object(values) => infer_object(values),
+    }
+}
+
+fn infer_bool(_value: &bool) -> Shape {
+    Shape {
+        type_: types::BOOLEAN,
+        ..Default::default()
+    }
+}
+
+fn infer_number(value: &serde_json::Number) -> Shape {
+    let type_ = if value.is_f64() {
+        types::FRACTIONAL
+    } else {
+        types::INTEGER
+    };
+
+    Shape {
+        type_,
+        ..Default::default()
+    }
+}
+
+fn infer_string(_value: &str) -> Shape {
+    Shape {
+        type_: types::STRING,
+        ..Default::default()
+    }
+}
+
+fn infer_null() -> Shape {
+    Shape {
+        type_: types::NULL,
+        ..Default::default()
+    }
+}
+
+fn infer_array(inner: &[JsonValue]) -> Shape {
+    if let Some(shape) = inner
+        .iter()
+        .map(infer_shape)
+        .reduce(|acc, v| shape::merge(acc, v))
+    {
+        Shape {
+            type_: types::ARRAY,
+            array: ArrayShape {
+                tuple: vec![shape],
```

Comment:
I think this use of `tuple` might not be right. The values in the array are all reduced into a single schema, so it seems like it should be set as `additional`, and leave `tuple` empty.